File: PriorityQueue/CPP/PriorityQueue.cpp
//
// Created by Shahriar Nasim Nafi on 7/7/22.
//

#include <iostream>

using namespace std;

const int MAX = 100;

struct MaxHeapify {
    static int left(int i) {
        return (i + 1) * 2 - 1;
    }

    static int right(int i) {
        return (i + 1) * 2;
    }

    static int parent(int i) {
        return (i + 1) / 2 - 1;
    }

    static void maxHeapify(int heap[], int heapSize, int root) {

        int largest = root;

        int left = MaxHeapify::left(root);
        int right = MaxHeapify::right(root);

        if (left <= heapSize && heap[left] > heap[largest]) {
            largest = left;
        }

        if (right <= heapSize && heap[right] > heap[largest]) {
            largest = right;
        }

        if (largest != root) {
            int temp = heap[root];
            heap[root] = heap[largest];
            heap[largest] = temp;

            maxHeapify(heap, heapSize, largest);
        }
    }

    static void buildHeap(int heap[], int heapSize) {
        for (int i = heapSize / 2 - 1; i >= 0; i--) {
            maxHeapify(heap, heapSize, i);
        }
    }
};


struct PriorityQueue {
public:
    static int extractMax(int *heap, int n) {
        int max = heap[0];
        heap[0] = heap[n - 1];
        n--;
        MaxHeapify::maxHeapify(heap, n - 1, 0);
        return max;
    }


    static void insert(int value, int *heap, int n) {
        heap[n] = value;
        int i = n;

        n++;

        int p = MaxHeapify::parent(i);
        while (i > 0 && heap[i] > heap[p]) {
            int temp = heap[i];
            heap[i] = heap[p];
            heap[p] = temp;
            i = p;
            p = MaxHeapify::parent(i);
        }
    }
};

void heapPrinter(int *heap, int heapSize, int root) {

    if (root > heapSize) {
        return;
    }

    cout << heap[root] << endl;
    int left = MaxHeapify::left(root);
    int right = MaxHeapify::right(root);


    if (heapSize >= left) {
        cout << heap[root] << " - " << heap[left] << endl;
    }

    if (heapSize >= right) {
        cout << heap[root] << " - " << heap[right] << endl;
    }

    heapPrinter(heap, heapSize, left);
    heapPrinter(heap, heapSize, right);

}

//
// Created by Shahriar Nasim Nafi on 7/7/22.
//

#include <iostream>

using namespace std;

const int MAX = 100;

struct MaxHeapify {
    static int left(int i) {
        return (i + 1) * 2 - 1;
    }

    static int right(int i) {
        return (i + 1) * 2;
    }

    static int parent(int i) {
        return (i + 1) / 2 - 1;
    }

    static void maxHeapify(int heap[], int heapSize, int root) {

        int largest = root;

        int left = MaxHeapify::left(root);
        int right = MaxHeapify::right(root);

        if (left <= heapSize && heap[left] > heap[largest]) {
            largest = left;
        }

        if (right <= heapSize && heap[right] > heap[largest]) {
            largest = right;
        }

        if (largest != root) {
            int temp = heap[root];
            heap[root] = heap[largest];
            heap[largest] = temp;

            maxHeapify(heap, heapSize, largest);
        }
    }

    static void buildHeap(int heap[], int heapSize) {
        for (int i = heapSize / 2 - 1; i >= 0; i--) {
            maxHeapify(heap, heapSize, i);
        }
    }
};


struct PriorityQueue {
public:
    static int extractMax(int *heap, int n) {
        int max = heap[0];
        heap[0] = heap[n - 1];
        n--;
        MaxHeapify::maxHeapify(heap, n - 1, 0);
        return max;
    }


    static void insert(int value, int *heap, int n) {
        heap[n] = value;
        int i = n;

        n++;

        int p = MaxHeapify::parent(i);
        while (i > 0 && heap[i] > heap[p]) {
            int temp = heap[i];
            heap[i] = heap[p];
            heap[p] = temp;
            i = p;
            p = MaxHeapify::parent(i);
        }
    }
};

void heapPrinter(int *heap, int heapSize, int root) {

    if (root > heapSize) {
        return;
    }

    cout << heap[root] << endl;
    int left = MaxHeapify::left(root);
    int right = MaxHeapify::right(root);


    if (heapSize >= left) {
        cout << heap[root] << " - " << heap[left] << endl;
    }

    if (heapSize >= right) {
        cout << heap[root] << " - " << heap[right] << endl;
    }

    heapPrinter(heap, heapSize, left);
    heapPrinter(heap, heapSize, right);

}

File: PriorityQueue/Swift/PriorityQueue.swift
//
// Created by Shahriar Nasim Nafi on 7/7/22.
//

struct MaxHeapify<T: Comparable> {
    
    static func leftChild(for index: Int) -> Int {
        (index + 1) * 2 - 1
    }
    
    static func rightChild(for index: Int) -> Int {
        (index + 1) * 2
    }
    
    static func parent(for index: Int) -> Int {
        (index + 1) / 2 - 1
    }
    
    static func maxHeapify(heap: inout [T], heapSize: Int, root: Int) {
        var largest = root
        let left = leftChild(for: root)
        let right = rightChild(for: root)
        
        if (left <= heapSize && heap[left] > heap[largest]) {
            largest = left
        }
        
        if (right <= heapSize && heap[right] > heap[largest]) {
            largest = right
        }
        
        if root != largest {
            (heap[root], heap[largest]) = (heap[largest], heap[root])
            maxHeapify(heap: &heap, heapSize: heapSize, root: largest)
        }
    }
    
    static func buildMaxHeap(heap: inout [T], heapSize: Int) {
        for i in stride(from: heapSize / 2 - 1 , through: 0, by: -1) {
            maxHeapify(heap: &heap, heapSize: heapSize, root: i)
        }
    }
    
    static func printHeap(heap: [T], heapSize: Int, root: Int) {
        if root > heapSize {
            return
        }
        
        print(heap[root])
        
        let left = leftChild(for: root)
        let right = rightChild(for: root)
        
        if left <= heapSize {
            print(heap[root], "->", heap[left])
        }
        
        if right <= heapSize {
            print(heap[root], "->", heap[right])
        }
        
        printHeap(heap: heap, heapSize: heapSize, root: left)
        printHeap(heap: heap, heapSize: heapSize, root: right)
    }
    
    static func heapSort(heap: inout [T], heapSize: Int) {
        for i in stride(from: heapSize, to: 0, by: -1) {
            (heap[i], heap[0]) = (heap[0], heap[i])
            maxHeapify(heap: &heap, heapSize: i - 1, root: 0)
        }
    }
    
    static func heapSorted(array: [T]) -> [T] {
        var sorted = array
        let heapSize =  array.count - 1
        
        buildMaxHeap(heap: &sorted, heapSize: heapSize)
        
        for i in stride(from: heapSize, to: 0, by: -1) {
            (sorted[i], sorted[0]) = (sorted[0], sorted[i])
            maxHeapify(heap: &sorted, heapSize: i - 1, root: 0)
        }
        
        return sorted
    }
    
}


struct PriorityQueue<T: Comparable> {
    static func extractMax(heap: inout [T]) -> T {
        let max = heap[0]
        
        heap[0] = heap[heap.count - 1]
        heap.removeLast()
        
        MaxHeapify.maxHeapify(heap: &heap, heapSize: heap.count - 1, root: 0)
        
        return max
    }
    
    static func insert(value: T, heap: inout [T]) {
        heap.append(value)
        
        var i = heap.count - 1
        var p = MaxHeapify<T>.parent(for: i)
        
        while i > 0 && heap[i] > heap[p] {
            (heap[i], heap[p]) = (heap[p], heap[i])
            i = p
            p = MaxHeapify<T>.parent(for: i)
        }
    }
}


